### ISSUE TITLE
Ignition spec 2.x partitioning is not translatable to tools other than sgdisk

### ISSUE THREAD

**Issue by ajeddeloh** [coreos/ignition#603]
# Bug #

## Operating System Version ##
Any

## Ignition Version ##
Spec version 2.x (maybe 1.x as well?)

## Environment ##
Any

## Overview of Partitioning Today:
The Ignition spec allows for specifying 0 for the "start" or "size" of a partition. This is passed through to sgdisk. It refers to the starting sector of the largest available block and the ending sector of the largest available block. This allows easy creation of a "last partition" filling the rest of the disk (start and size both 0) and creating a bunch of new partitions on the same disk (start 0 and defined size), but is disastrous in all other situations. For one, size 0 always means "try to make the partition end at the end of the largest block" regardless of what start is. This is unintuitive and not useful. (A more useful version would be to make the partition as big as possible with the specified start).

We want to change the partitioning support with 3.0 to break from these semantics and either eliminate them (as an attractive nuisance) or replace them with something more intuitive (like have size 0 mean "as big as possible for the given starting sector"). Ignition internally translates old configs to the latest.

The problem is that the meaning of "0" in old configs is dependent on the contents of the disk it's defined for; it has no concrete meaning by itself without the context of the contents of the disk. If the new config fixes the quirks of the old version, there's no way to translate it to the new version since nothing with the same semantics exists in the new version and we don't know the contents of the disk it will be run on.

This leaves four awful options:
1) Keep the sgdisk semantics forever in the spec
2) Change the spec. Have some future Ignition fail if it encounters 0's for start/size in a 2.x spec since it can't translate it to a new spec version with the same meaning.
3) Change the spec. Have some mechanism for detecting this config uses "tainted" start/size 0 and carry twice the partitioning logic in the disks stage.
4) Change the spec. Have the translator make a "best effort" translation.

**Comment by crawford** [comment]
There is also option 5: Change the spec. Use the version to determine whether sgdisk or some other tool will be used.

**Comment by ajeddeloh** [comment]
That's option 3.

**Comment by ajeddeloh** [comment]
After some further thought, this isn't as bad as it seems. We are still tied into supporting the semantics of 2.x configs but they could be a subset of newer semantics.

I'm going to use the terms EOLB and SOLB for end/start of largest block.

For size=0 (currently means "sized such that it ends at the EOLB") there are three cases:
1) The current block (starting at the provided starting sector) ends at EOLB. In this case the current "size 0" is the same as filling the rest of the block. I'd argue this is what we want. 
2) The current block ends after EOLB. This must means that the starting sector is not in the largest block at all (otherwise this would be the same as case 1). This also means the EOLB is before the starting sector, so Ignition will error out since the end is before the start.
3) The current block ends before the EOLB. This also must mean the starting sector is not in the largest block (or else it would be case 1). This means there is an existing partition between the starting sector and the ELOB. Ignition will error since there are overlapping partitions.

Start=0 is a little simpler since the Ignition spec specifies a size and not and ending sector. There are two good canidates for what start=0 should mean:
1) The start of the largest block (current meaning)
2) The start of the first available block.

#2 is more intuitive but after some thought, I think sgdisk actually does the right thing by picking #1. #1 allows you to ensure in your Ignition config the earlier partitions get the largest blocks.

Ultimately start=0 isn't even that big an issue since (I think) most people will be using it in the case of a blank disk where #1 and #2's semantics end up with the same result.

In summary: size=0 isn't an issue because the cases where it doesn't do what we want it just fails. I think the current start=0 semantics are good and we shouldn't change them.

**Comment by ajeddeloh** [comment]
Given the above, I'm going to close this unless anyone disagrees that
1) size 0 should mean "fill as much space as possible for the given starting sector"
2) start 0 should mean "the start of the largest available block.

Feel free to reopen if you feel otherwise.
